File: lista/GO/questao7/questao7.go
package main

import (
	"fmt"
)

type Escola struct {
	Nome        string
	Professores []*Professor
}

func (e *Escola) AdicionarProfessor(p *Professor) {
	e.Professores = append(e.Professores, p)
}

func (e Escola) ExibirProfessores() {
	fmt.Printf("Escola: %s\n", e.Nome)
	fmt.Println("Professores:")
	for _, professor := range e.Professores {
		fmt.Printf("° %s\n", professor.Nome)
	}
}

type Professor struct {
	Nome    string
	Escolas []*Escola
}

func (p *Professor) AdicionarEscola(e *Escola) {
	p.Escolas = append(p.Escolas, e)
}

func (p Professor) ExibirEscolas() {
	fmt.Printf("Professor: %s\n", p.Nome)
	fmt.Println("Escolas:")
	for _, escola := range p.Escolas {
		fmt.Printf("° %s\n", escola.Nome)
	}
}

func main() {
	escola1 := Escola{Nome: "Escola A"}
	escola2 := Escola{Nome: "Escola B"}

	professor1 := Professor{Nome: "Maria"}
	professor2 := Professor{Nome: "João"}

	escola1.AdicionarProfessor(&professor1)
	escola1.AdicionarProfessor(&professor2)
	escola2.AdicionarProfessor(&professor2)

	professor1.AdicionarEscola(&escola1)
	professor2.AdicionarEscola(&escola1)
	professor2.AdicionarEscola(&escola2)

	escola1.ExibirProfessores()
	escola2.ExibirProfessores()

	professor1.ExibirEscolas()
	professor2.ExibirEscolas()
}

File: lista/Python/questao7.py
"""
Crie uma classe Escola e uma classe Professor. A associação deve permitir
que uma escola tenha vários professores, e um professor possa lecionar em várias escolas.
"""
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.adicionar_professor(self)

    def exibir_escolas(self):
        print(f"Professor {self.nome} ensina nas seguintes escolas:")
        for escola in self.escolas:
            print(f"° {escola.nome}")

class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []
    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.adicionar_escola(self)

    def exibir_professores(self):
        print(f"A escola {self.nome} tem os seguintes professores:")
        for professor in self.professores:
            print(f"° {professor.nome}")

def main():
    escola1 = Escola("Maria Caxias")
    escola2 = Escola("João da Silva")

    professor1 = Professor("João")
    professor2 = Professor("Maria")
    professor3 = Professor("José")

    escola1.adicionar_professor(professor1)
    escola1.adicionar_professor(professor2)
    escola1.adicionar_professor(professor3)
    escola2.adicionar_professor(professor1)
    escola2.adicionar_professor(professor3)


    professor1.exibir_escolas()
    professor2.exibir_escolas()
    professor3.exibir_escolas()

    escola1.exibir_professores()
    escola2.exibir_professores()

if __name__ == "__main__":
    main()
